File: src/main/java/ru/dargen/evoplus/event/EventBus.java
package ru.dargen.evoplus.event;

import lombok.val;
import ru.dargen.evoplus.EvoPlus;

import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

public final class EventBus {

    private final Map<Class<? extends Event>, Set<Consumer<Event>>> handlerLists = new ConcurrentHashMap<>();
    private final EvoPlus mod;

    public EventBus(EvoPlus mod) {
        this.mod = mod;
    }

    public <E extends Event> void register(Class<E> eventClass, Consumer<E> handler) {
        handlerLists.computeIfAbsent(eventClass, (__) -> new HashSet<>()).add((Consumer<Event>) handler);
    }

    public <E extends Event> Set<Consumer<Event>> getHandlerList(Class<E> eventClass) {
        return handlerLists.getOrDefault(eventClass, Collections.emptySet());
    }

    public <E extends Event> E fireEvent(E event) {
        val eventClass = event.getClass();
        val handlerList = getHandlerList(eventClass);

        handlerList.forEach(handler -> {
            try {
                handler.accept(event);
            } catch (Throwable t) {
                mod.getLogger().error("Error while handle event " + eventClass.getName(), t);
            }
        });

        return event;
    }

}

File: src/main/kotlin/ru/dargen/evoplus/event/EventBus.kt
package ru.dargen.evoplus.event

import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap
import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet
import ru.dargen.evoplus.Logger
import ru.dargen.evoplus.util.kotlin.cast
import java.util.logging.Level

typealias EventHandler<E> = E.() -> Unit

object EventBus {

    private val Handlers: MutableMap<Class<*>, MutableSet<EventHandler<*>>> = Object2ObjectOpenHashMap()

    fun <E : Event> register(type: Class<E>, handler: EventHandler<E>) =
        Handlers.getOrPut(type, ::ObjectOpenHashSet).add(handler.cast())

    fun <E : Event> fire(event: E) = event.apply {
        Handlers[event.javaClass]?.forEach { handler ->
            runCatching { handler.cast<EventHandler<E>>()(event) }
                .exceptionOrNull()
                ?.let { Logger.log(Level.SEVERE, "Error while event dispatch ${event.javaClass}", it) }
        }
    }

    fun <E : CancellableEvent> fireResult(event: E) = !fire(event).isCancelled

}

inline fun <reified E : Event> on(noinline handler: EventHandler<E>) = EventBus.register(E::class.java, handler)